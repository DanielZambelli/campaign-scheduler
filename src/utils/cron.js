const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
