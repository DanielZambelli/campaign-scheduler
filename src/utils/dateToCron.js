const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
