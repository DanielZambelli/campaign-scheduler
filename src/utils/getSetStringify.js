const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
