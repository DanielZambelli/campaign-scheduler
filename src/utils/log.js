const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
