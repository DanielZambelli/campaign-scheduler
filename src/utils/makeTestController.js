const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
