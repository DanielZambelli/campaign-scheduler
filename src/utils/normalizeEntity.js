const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
