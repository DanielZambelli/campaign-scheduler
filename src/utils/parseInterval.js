const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
