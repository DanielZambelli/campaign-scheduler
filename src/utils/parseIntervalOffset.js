const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
