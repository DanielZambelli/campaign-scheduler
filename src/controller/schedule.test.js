require('../utils/test')
const {seed} = require('../utils/test/seeds1')
const TEST = 'schedule'

describe(TEST, () => {

  beforeEach(init(TEST, seed))

  afterEach(destroy)

  it('validates', async () => {
    const res = [
      await ctl.schedule('campaign1', undefined).catch(e => e.message),
      await ctl.schedule('campaign1', null).catch(e => e.message),
      await ctl.schedule('campaign1', 0).catch(e => e.message),
      await ctl.schedule(undefined, undefined).catch(e => e.message),
    ]
    expect(res).toEqual([
      'requires campaignId and subject',
      'requires campaignId and subject',
      'requires campaignId and subject',
      'requires campaignId and subject'
    ])
  })

  it('validates', async () => {
    const res = await ctl.schedule('NON_EXISTING_CAMPAIGN', 'contact#1').catch(e => e.message)
    expect(res).toEqual('campaign not found')
  })

  it('returns', async () => {
    const res = await ctl.schedule('campaign1', 'contact#1').then(normalize)
    expect(res.result.schedule.active).toEqual(true)
    expect(res.result.actions.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('returns', async () => {
    const res = await ctl.schedule('campaign2', 'contact#1').then(normalize)
    expect(res.result.schedule.active).toEqual(true)
    expect(res.result.actions.length).toEqual(0)
    expect(res).toMatchSnapshot()
  })

  it('returns', async () => {
    const res = await ctl.schedule('campaign3', 'contact#1').then(normalize)
    expect(res.result.schedule.active).toEqual(true)
    expect(res.result.actions.length).toEqual(2)
    expect(res).toMatchSnapshot()
  })

  it('returns', async () => {
    const res = normalize([
      await ctl.schedule('campaign1', 'contact#1'),
      await ctl.schedule('campaign2', 'contact#2'),
      await ctl.schedule('campaign3', 'contact#3'),
      await ctl.schedule('campaign1', 'contact#4'),
    ])
    expect(res.result[0].schedule.active).toEqual(true)
    expect(res.result[0].actions.length).toEqual(1)
    expect(res.result[1].schedule.active).toEqual(true)
    expect(res.result[1].actions.length).toEqual(0)
    expect(res.result[2].schedule.active).toEqual(true)
    expect(res.result[2].actions.length).toEqual(2)
    expect(res.result[3].schedule.active).toEqual(true)
    expect(res.result[3].actions.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    await ctl.schedule('campaign1', 'contact#1')
    const res = await getState({ campaignId: 'campaign1' })
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    await ctl.schedule('campaign1', 'contact#2')
    const res = await getState({ campaignId: 'campaign1' })
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    await ctl.schedule('campaign2', 'contact#1')
    const res = await getState({ campaignId: 'campaign2' })
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    await ctl.schedule('campaign3', 'contact#1')
    const res = await getState()
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    await ctl.schedule('campaign1', 'contact#1')
    await ctl.schedule('campaign1', 'contact#2')
    await ctl.schedule('campaign1', 'contact#3')
    await ctl.schedule('campaign1', 'contact#4')
    await ctl.schedule('campaign1', 'portal#1')
    await ctl.schedule('campaign1', 'portal#2')
    await ctl.schedule('campaign2', 'contact#1')
    await ctl.schedule('campaign2', 'contact#2')
    await ctl.schedule('campaign3', 'contact#3')
    await ctl.schedule('campaign3', 'contact#4')
    const res = await getState()
    expect(res).toMatchSnapshot()
  })

  it('avoid duplicates', async () => {
    await ctl.schedule('campaign3', 'contact#1')
    await ctl.schedule('campaign3', 'contact#1')
    await ctl.schedule('campaign3', 'contact#1')
    await ctl.schedule('campaign3', 'contact#1')
    const res = await getState()
    expect(res).toMatchSnapshot()
  })

  it('reschedules pending', async () => {
    await ctl.schedule('campaign1', 'contact#1')
    await ctl.db.Schedules.update({ active: false }, { where: { }})
    await ctl.db.Actions.destroy({ where: { }})
    await ctl.schedule('campaign1', 'contact#1')
    const res = await getState({ campaignId: 'campaign1' })
    expect(res.state.schedules[0].active).toEqual(true)
    expect(res.state.actions.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('reschedules completed', async () => {
    await ctl.schedule('campaign3', 'contact#1')
    await ctl.db.Schedules.update({ active: false }, { where: { }})
    await ctl.db.Actions.update({ state: 'completed' }, { where: { actionId: 'action1' }})
    await ctl.db.Actions.destroy({ where: { state: 'pending' }})
    await ctl.schedule('campaign3', 'contact#1')
    const res = await getState({ campaignId: 'campaign3' })
    expect(res.state.schedules[0].active).toEqual(true)
    expect(res.state.actions.length).toEqual(2)
    expect(res.state.actions[0].state).toEqual('completed')
    expect(res.state.actions[1].state).toEqual('pending')
    expect(res).toMatchSnapshot()
  })

})
