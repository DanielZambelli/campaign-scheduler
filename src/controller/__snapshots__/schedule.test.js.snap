// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedule avoid duplicates 1`] = `
{
  "hash": "4f7805139150b40ab4d41f5efdb2e4891f879567",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`schedule reschedules completed 1`] = `
{
  "hash": "6120edd36a60e7b1829fa1fec8f249f2a627f584",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`schedule reschedules pending 1`] = `
{
  "hash": "634e8daf5f329858636158380bea77ad48953874",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`schedule returns 1`] = `
{
  "hash": "5c8114317dc6074ef6b5d7e3d4a0b6e18ddabe22",
  "result": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "createdAt": "a few seconds ago",
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
        "updatedAt": "a few seconds ago",
      },
    ],
    "schedule": {
      "active": true,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
  },
}
`;

exports[`schedule returns 2`] = `
{
  "hash": "bcc80dd45cfad716d96ce9ef376c58bbeeee0d6b",
  "result": {
    "actions": [],
    "schedule": {
      "active": true,
      "campaignId": "campaign2",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
  },
}
`;

exports[`schedule returns 3`] = `
{
  "hash": "34a56ee0c62ef609af7b886b319b51a93ad58cd0",
  "result": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "createdAt": "a few seconds ago",
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
        "updatedAt": "a few seconds ago",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "createdAt": "a few seconds ago",
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
        "updatedAt": "a few seconds ago",
      },
    ],
    "schedule": {
      "active": true,
      "campaignId": "campaign3",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
  },
}
`;

exports[`schedule returns 4`] = `
{
  "hash": "f235337f328944d817a0a5c03061b6ee29838c03",
  "result": [
    {
      "actions": [
        {
          "actionId": "action1",
          "callbackId": "sendEmail",
          "callbackOpts": {
            "view": "emailTemplate1",
          },
          "campaignId": "campaign1",
          "completedAt": null,
          "createdAt": "a few seconds ago",
          "expectedAt": "in an hour",
          "log": null,
          "state": "pending",
          "subject": "contact#1",
          "updatedAt": "a few seconds ago",
        },
      ],
      "schedule": {
        "active": true,
        "campaignId": "campaign1",
        "createdAt": "a few seconds ago",
        "subject": "contact#1",
        "updatedAt": "a few seconds ago",
      },
    },
    {
      "actions": [],
      "schedule": {
        "active": true,
        "campaignId": "campaign2",
        "createdAt": "a few seconds ago",
        "subject": "contact#2",
        "updatedAt": "a few seconds ago",
      },
    },
    {
      "actions": [
        {
          "actionId": "action1",
          "callbackId": "sendEmail",
          "callbackOpts": {
            "view": "emailTemplate1",
          },
          "campaignId": "campaign3",
          "completedAt": null,
          "createdAt": "a few seconds ago",
          "expectedAt": "in a day",
          "log": null,
          "state": "pending",
          "subject": "contact#3",
          "updatedAt": "a few seconds ago",
        },
        {
          "actionId": "action2",
          "callbackId": "sendEmail",
          "callbackOpts": {
            "view": "emailTemplate1",
          },
          "campaignId": "campaign3",
          "completedAt": null,
          "createdAt": "a few seconds ago",
          "expectedAt": "in 2 days",
          "log": null,
          "state": "pending",
          "subject": "contact#3",
          "updatedAt": "a few seconds ago",
        },
      ],
      "schedule": {
        "active": true,
        "campaignId": "campaign3",
        "createdAt": "a few seconds ago",
        "subject": "contact#3",
        "updatedAt": "a few seconds ago",
      },
    },
    {
      "actions": [
        {
          "actionId": "action1",
          "callbackId": "sendEmail",
          "callbackOpts": {
            "view": "emailTemplate1",
          },
          "campaignId": "campaign1",
          "completedAt": null,
          "createdAt": "a few seconds ago",
          "expectedAt": "in an hour",
          "log": null,
          "state": "pending",
          "subject": "contact#4",
          "updatedAt": "a few seconds ago",
        },
      ],
      "schedule": {
        "active": true,
        "campaignId": "campaign1",
        "createdAt": "a few seconds ago",
        "subject": "contact#4",
        "updatedAt": "a few seconds ago",
      },
    },
  ],
}
`;

exports[`schedule state 1`] = `
{
  "hash": "634e8daf5f329858636158380bea77ad48953874",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`schedule state 2`] = `
{
  "hash": "bf05003bc4c21683757b5b9c4ed9111051adda10",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#2",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#2",
      },
    ],
  },
}
`;

exports[`schedule state 3`] = `
{
  "hash": "c77246581dd3be1f05714d9e3e63a6a7b37c08a5",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`schedule state 4`] = `
{
  "hash": "4f7805139150b40ab4d41f5efdb2e4891f879567",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`schedule state 5`] = `
{
  "hash": "1d877a34838593a379dae304d62a4dc16fb096ac",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#2",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#3",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "contact#4",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "portal#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "pending",
        "subject": "portal#2",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#3",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#3",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#4",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#4",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#2",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#3",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#4",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "portal#1",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "portal#2",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#2",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#3",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#4",
      },
    ],
  },
}
`;
